### PR TITLE
Rewrite the balancing algorithm so that tx is always pre-balanced before ex units calculation

### PR DESCRIPTION
**Pull request by errfrom**
Closes https://github.com/Plutonomicon/cardano-transaction-lib/issues/892, Closes https://github.com/Plutonomicon/cardano-transaction-lib/issues/570 (we have a similar problem with CSL fee calculation) 

TODO: description

### Pre-review checklist

- [x] All code has been formatted using our config (`make format`)
- [x] The integration and unit tests have been run (`npm run test`) locally
- [ ] The changelog has been updated under the `## Unreleased` header, using the appropriate sub-headings (`### Added`, `### Removed`, `### Fixed`), and the links to the appropriate issues/PRs have been included

### CODE REVIEW

**Review by ngua — CHANGES_REQUESTED** [review]
Excellent work! AFAICT the new, simplified balancer is working correctly and I am very much in favor of the changes you introduce here

My request for changes is not related to your work. I just think this is good opportunity to further cut down on some earlier technical debt in CTL

File: src/BalanceTx/Error.purs [hunk 1]
```diff
@@ -0,0 +1,218 @@
+module BalanceTx.Error
```
Comment (ngua) [inline src/BalanceTx/Error.purs:1]: Could use a module doc comment

**Review by ngua — COMMENTED** [review]
Great work! Just a couple of final things. I can take care of these now to merge this in more quickly

**Review by ngua — APPROVED** [review]

**Comment by errfrom** [comment]
@ngua Thanks for your review! I would like to make some more changes real quick, and then we can merge it

**Review by ngua — APPROVED** [review]

**Pull request merged**